### ISSUE TITLE
[RFC] Delta weight sync on Ascend NPU (HCCL transport)

### ISSUE THREAD

**Issue by HaoWang113** [verl-project/verl#7787]
Draft implementation on my fork: https://github.com/HaoWang113/verl/pull/1 (20 files, +2554/−627, six commits), extending the delta weight sync roadmap (#7060, #6974, #7227) to Ascend NPU.

> **Status:** the HCCL transport is functionally complete, but the code is still being polished and the NPU test ladder / benchmarks are running in parallel. **This issue is not a request for code review** — we are looking for early feedback on the overall design and on the open questions at the bottom. Test results and benchmark data will be attached to the draft PR before it is submitted upstream.

## Motivation

verl's delta weight sync (`delta_sharded` checkpoint engine) lets the trainer skip unchanged parameters when syncing weights into the rollout engine, but it is currently CUDA-only: the NCCL delta engine requires `cupy` and forwards flushes over CUDA IPC, neither of which exists on Ascend NPU hosts. NPU training runs therefore pay a **full-weight resync on every weight update** — for reference, the dense HCCL engine (#4885) costs ~11 s per full sync at 5.3 GB/s on 2×16 Ascend 910C with Qwen3-30B-A3B.

On the vllm-ascend side there is a community sparse weight-transfer effort (vllm-ascend#13682), but it uses runtime parameter coordinates with TP=1/PP=1 and per-patch broadcasts, so it is not consumable by verl's checkpoint-coordinate pipeline (details under *Questions*).

## Proposal

The design rests on one observation: the delta-sync pipeline is transport-agnostic at its core — the state machine, bucketing and manifest publishing do not care how bytes reach the receiver — so the refactor extracts that core and lets each transport implement only the wire.

```yaml
actor_rollout_ref:
  rollout:
    checkpoint_engine:
      backend: delta_sharded   # identical config on NPU and CUDA hosts
```

- **Zero new config for NPU users.** `delta_sharded` resolves to the HCCL delta engine on NPU and to the NCCL engine on CUDA (import-arbitrated: the NCCL module needs `cupy` and fails to import on NPU). The vLLM weight-transfer backend likewise defaults to `verl_delta_hccl` on NPU and `verl_delta_ipc` elsewhere — configs stay portable across platforms.
- **One pipeline, two transports.** The seed/steady state machine, bucketing, flush manifests and wire schema are extracted into a transport-agnostic `DeltaSyncMixin` shared by the NCCL and HCCL engines. The CUDA path is refactored but behaviorally unchanged (guarded by the existing GPU suites).
- **Direct broadcast into the vLLM workers.** NPU has no CUDA IPC, so flush payloads are not forwarded through the rollout adapter. Instead the trainer's rank 0 broadcasts each flush into a PyHccl broadcast group **whose members are the vLLM worker processes themselves** — one zmq manifest subscription plus two broadcasts per flush, no IPC-handle negotiation, persistent group.
- **Device-agnostic checksum.** The wire checksum moves from `torch.hash_tensor` (PyTorch ≥ 2.9, CPU/CUDA kernels only) to a device-side XOR tree — one implementation that runs on any device (see *Questions* #3).

```
trainer side                                    rollout side (vLLM worker process)
─────────────                                   ─────────────────────────────────
FSDP shard export (unchanged, #7227 contract)
        │
DeltaSyncMixin (delta_sync/pipeline.py)   ←  shared state machine, bucketing,
        │                                      zmq manifest publishing
        ├── NCCL engine → CUDA IPC → vLLM / SGLang
        └── HCCL engine (NPU)
                │ trainer rank 0                VerlDeltaHCCLWeightTransferEngine
                │  PyHccl broadcast group ────▶ (joins the group inside the
                │  (members = the vLLM workers)    worker process)
```

### The abstraction

Today every delta-sync concern lives inside the NCCL engine, with CUDA-IPC transport details interleaved. The refactor pulls the transport-agnostic core into `DeltaSyncMixin` (`verl/checkpoint_engine/delta_sync/pipeline.py`), which owns:

- the **seed/steady state machine** — the first sync streams the backend's full per-tensor export (doubling as the full-sync baseline), later syncs gather only changed elements per parameter slot;
- **bucketing** — a `_GatherQueue` collects per-slot deltas on rank 0, and `_FlushBucket` accumulates them into size-capped flushes, each sealed into one `DeltaFlush` wire message;
- **streaming + manifest publishing** — each flush goes on the wire as soon as it is sealed (trainer peak memory stays near two buckets, even during the seed); a zmq manifest announces each flush and carries the `is_last` marker.

Each transport then implements exactly four hooks:

| hook | contract |
|---|---|
| `_wire_device` | device the flush tensors are assembled and broadcast from (`cuda` / `npu`) |
| `_ensure_wire` | lazy transport init, called **before** each flush's manifest is published |
| `_broadcast_flush` | put one flush's raw uint8 buffers on the wire: `(pos_u8, val_u8)` — positions absent for values-only flushes (the seed) |
| `_release_after_round` | return transport staging memory after each round (cupy pool on CUDA; `npu.empty_cache` on NPU) |

The NCCL engine shrinks to these four (CUDA behavior unchanged, guarded by the existing GPU suites); the HCCL engine implements them over a persistent PyHccl broadcast group. The `_ensure_wire` ordering is deliberate: the HCCL group completes only once every receiver has joined, and every receiver connects its zmq SUB before joining — so *group ready ⇒ subscribed*, which is what makes publishing the manifest after this hook lossless despite zmq PUB's drop-before-subscribe semantics.

A second, orthogonal extraction: `delta_sync.decode` (`decode_delta_wire`) validates one received flush — spec JSON, encoding consistency, per-parameter slice bounds, checksum — and returns plain `DeltaPatch` entries on any device; the consumer only converts them into engine-specific patches. Together with `delta_sync/hccl_receiver.py` (shared zmq SUB, communicator builder and session pump), the same receive stack serves both the E2E test harness and the vLLM weight-transfer backend.

## Scope of the first revision

Supported on NPU: vLLM rollout only, BF16 non-quantized **dense** models, TP ∈ {1, 2} (covered by the E2E test), `data_parallel_size=1`, `pipeline_parallel_size=1`, `verify_every=0`.

Rejected with explicit startup errors, to be lifted in follow-ups: MoE models (no NPU MoE kernel path in vllm-ascend yet — roadmap follow-up), `rebuild_group=True` (the HCCL group is persistent; group setup dominates), the SGLang receiver (rides the CUDA IPC path).

Also out of scope for the first revision: worker crash/restart mid-session — the persistent group and the SUB-before-join rendezvous assume stable membership within a session (follow-up). If dense-first is a concern for your NPU workloads, speak up on this thread.

## Cost

Benchmark data is being collected on an 8× Ascend NPU host with Qwen3-8B-Base BF16 (seed round = full-sync baseline; steady rounds at 0/1/10/25/50/100 % sparsity, medians over 3 rounds) and will be attached to the draft PR. For orientation, the dense HCCL full sync costs ~11 s on 2×16 910C / Qwen3-30B-A3B; a delta round should scale with the changed ratio rather than model size.

Per round the transported payload is the changed elements only (flush bucketing is exercised at 3072 MB and 128 MB in tests); every sync reports `changed_ratio` / `payload_mbytes` / `flushes` / `elapsed_seconds` metrics from `CheckpointEngineManager.update_weights`.

## Landing plan

Currently one draft PR of six independently-reviewable commits — mixin extraction → decode extraction → HCCL engine → vLLM weight-transfer backend → tests → docs (per-commit breakdown in the PR). Happy to regroup into separate PRs (mixin refactor / HCCL engine / vLLM backend) if reviewers prefer smaller increments.

## Questions for maintainers

1. **Receiver geometry.** NPU has no CUDA IPC, so flushes are broadcast directly into the vLLM worker processes (they join the PyHccl group themselves) instead of being forwarded through the rollout adapter. This removes a copy and a hop but couples worker processes to the trainer's group — feedback welcome, especially from maintainers of the disaggregated-rollout paths.
2. **Backend naming.** `delta_sharded` resolves to the HCCL engine on NPU and the NCCL engine on CUDA (import-arbitrated, zero-config, portable configs). Alternative: an explicit `delta_sharded_hccl` name for discoverability at the cost of platform-conditional configs. Which do you prefer?
3. **Checksum change affects the NCCL wire too.** The XOR-tree checksum replaces `torch.hash_tensor` for all transports (one implementation, any device). Both endpoints switch together in the PR, but a version-skewed sender/receiver pair will fail the checksum rather than silently mismatch — acceptable, or should the checksum scheme be negotiated via the manifest?
4. **Relationship to vllm-ascend#13682.** That PR broadcasts per-patch indices/values with runtime coordinates and TP=1/PP=1 (fixed overhead dominates its reported numbers); ours buckets many parameters per flush and applies checkpoint-coordinate patches with TP>1. Once it lands we would like to converge the two HCCL rendezvous stacks — cc vllm-ascend maintainers.
